### PR TITLE
feat(ioctl): support znode project management commands(based #3987)

### LINKED ISSUE #3982: Introduce wsctl

### PR DESCRIPTION
**Pull request by saitofun**
# Description

support project create, update and query command

Fixes #3982 

## Type of change
Please delete options that are not relevant.
- [x] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration
- [x] make test

# Checklist:
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: ioctl/cmd/ws/wsproject.go [hunk 1]
```diff
@@ -0,0 +1,296 @@
...
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"math/big"
+	"reflect"
+	"strings"
+	"time"
+
+	"github.com/ethereum/go-ethereum/accounts/abi"
+	"github.com/iotexproject/iotex-core/ioctl/config"
+	"github.com/spf13/cobra"
```
Comment (CoderZhi) [inline ioctl/cmd/ws/wsproject.go:29]: org

File: ioctl/cmd/ws/wsprojectcreate.go [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+	_flagProjectHashUsages = map[config.Language]string{
+		config.English: "project config hash for validating",
+		config.Chinese: "项目配置hash",
+	}
+)
+
+func init() {
+	wsProjectCreate.Flags().StringP("project-uri", "u", "", config.TranslateInLang(_flagProjectUriUsages, config.UILanguage))
+	wsProjectCreate.Flags().StringP("project-hash", "v", "", config.TranslateInLang(_flagProjectHashUsages, config.UILanguage))
+	wsProjectCreate.Flags().StringP("contract-address", "v", "", config.TranslateInLang(_flagProjectHashUsages, config.UILanguage))
```
Comment (CoderZhi) [inline ioctl/cmd/ws/wsprojectcreate.go:56]: both are 'v'?

File: ioctl/cmd/ws/wsprojectcreate.go [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+	}
+
+	hashArg, err := convertStringToAbiBytes32(hash)
+	if err != nil {
+		return "", err
+	}
+
+	bytecode, err := wsProjectRegisterContractABI.Pack(
+		createWsProjectFuncName,
+		uri, hashArg,
```
Comment (CoderZhi) [inline ioctl/cmd/ws/wsprojectcreate.go:75]: new line

File: ioctl/cmd/ws/wsproject.go [hunk 4]
```diff
@@ -0,0 +1,298 @@
...
+
+	_flagProjectRegisterContractAddressUsages = map[config.Language]string{
+		config.English: "project register contract address",
+		config.Chinese: "项目注册合约地址",
+	}
+
+	wsProjectRegisterContractAddress string
+	wsProjectRegisterContractABI     abi.ABI
+
+	//go:embed wsproject.json
```
Comment (envestcc) [inline ioctl/cmd/ws/wsproject.go:54]: Learned `embed` from it!

File: ioctl/cmd/ws/wsproject.go [hunk 5]
```diff
@@ -0,0 +1,298 @@
...
+	wsProjectRegisterContractJSONABI []byte
+	_                                = embed.FS{}
```
Comment (envestcc) [inline ioctl/cmd/ws/wsproject.go:56]: Does this line aims to add dependency for embed pkg?

File: ioctl/cmd/action/action.go [hunk 6]
```diff
@@ -212,26 +212,6 @@ func fixGasLimit(caller string, execution *action.Execution) (*action.Execution,
 
 // SendRaw sends raw action to blockchain
 func SendRaw(selp *iotextypes.Action) error {
-	conn, err := util.ConnectToEndpoint(config.ReadConfig.SecureConnect && !config.Insecure)
```
[the remark this answers is not in the document]
Comment (saitofun) [reply ioctl/cmd/action/action.go:215]: `SendRaw` is still used elsewhere.

File: ioctl/cmd/ws/wsproject.go [hunk 7]
```diff
@@ -0,0 +1,298 @@
...
+		"contract-address",
+		"c",
+		"",
+		config.TranslateInLang(_flagProjectRegisterContractAddressUsages, config.UILanguage),
+	)
+	_ = wsProject.MarkFlagRequired("contract-address")
+}
+
+func convertStringToAbiBytes32(hash string) (interface{}, error) {
+	t, _ := abi.NewType("bytes32", "", nil)
```
Comment (envestcc) [inline ioctl/cmd/ws/wsproject.go:87]: Maybe  using `[32]byte` is enough

File: ioctl/cmd/ws/wsproject.go [hunk 8]
```diff
@@ -0,0 +1,298 @@
...
+		ctx = metautils.NiceMD(jwtMD).ToOutgoing(ctx)
+	}
+
+	var rsp *iotexapi.GetReceiptByActionResponse
+	err = backoff.Retry(func() error {
+		rsp, err = cli.GetReceiptByAction(ctx, &iotexapi.GetReceiptByActionRequest{
+			ActionHash: h,
+		})
+		return err
+	}, backoff.WithMaxRetries(backoff.NewConstantBackOff(30*time.Second), 3))
```
[the remark this answers is not in the document]
Comment (saitofun) [reply ioctl/cmd/ws/wsproject.go:128]: max retry 3 times, backoff 30 seconds.

File: ioctl/cmd/ws/wsproject.go [hunk 9]
```diff
@@ -0,0 +1,298 @@
...
+			return nil, output.NewError(output.APIError, "not found", nil)
+		} else if ok {
+			return nil, output.NewError(output.APIError, sta.Message(), nil)
+		}
+		return nil, output.NewError(output.NetworkError, "failed to invoke GetReceiptByAction api", err)
+	}
+	return rsp, nil
+}
+
+func getEventInputsByName(logs []*iotextypes.Log, eventName string) (map[string]any, error) {
```
Comment (envestcc) [inline ioctl/cmd/ws/wsproject.go:141]: It's similar to `unpackEventParam` in blockindex/contractstaking/util.go, can we make a common util function?

File: ioctl/cmd/ws/wsproject.go [hunk 10]
```diff
@@ -0,0 +1,298 @@
...
+		}
+	}
+	if err := abi.ParseTopicsIntoMap(inputs, args, topics); err != nil {
+		return nil, errors.Wrap(err, "unpack event indexed fields failed")
+	}
+
+	return inputs, nil
+}
+
+func parseOutput(targetAbi *abi.ABI, targetMethod string, result string) (string, error) {
```
Comment (envestcc) [inline ioctl/cmd/ws/wsproject.go:186]: NOT duplicate the `parseOutput` func from ioctl/cmd/contract/parse.go

File: ioctl/cmd/action/action.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (envestcc) [reply ioctl/cmd/action/action.go:215]: I mean `SendRaw` doesn't actually send data anymore

**Review by envestcc — APPROVED** [review]

**Review by CoderZhi — APPROVED** [review]

**Pull request merged**